lib/puppet/provider/foreman_resource/rest_v3.rb

```diff
@@ -96,6 +99,20 @@ def success?(response)
   end
 
   def error_message(response)
-    JSON.parse(response.body)['error']['full_messages'].join(' ') rescue "unknown error (response #{response.code})"
+    explanations = {
+      '400' => 'Something is wrong with the data sent to Foreman server',
+      '401' => 'Often this is caused by invalid Oauth credentials',
+      '404' => 'The requested resource was not found',
+      '500' => 'Check /var/log/foreman/production.log on Foreman server for detailed information',
```

Comment:
Since [#28384](https://projects.theforeman.org/issues/28384) (Foreman 2.2) there is a rake task. If we have the request ID (available as the `X-Request-Id` header) we can recommend `REQUEST_ID=... foreman-rake errors:fetch_log` instead of the precise error. I don't have experience with this myself, but if we want to make it easy to debug we can be much more precise nowadays. This task has to run on the Foreman server but the registration can happen from any server so we can't directly run it.

Note that Apache can also return a HTTP 500 in which case you need to look in the Apache logs. That was more common with Passenger (for example, when Rails failed to even boot) but may still happen.